local_ads/icons_info.cpp

```diff
@@ -0,0 +1,69 @@
+#include "local_ads/icons_info.hpp"
+
+#include "coding/reader.hpp"
+
+#include "platform/platform.hpp"
+
+#include "base/assert.hpp"
+#include "base/logging.hpp"
+#include "base/string_utils.hpp"
+
+#include <functional>
+
+namespace
+{
+char const kDelimiter = '_';
+
+void ParseIconsFile(std::string const & iconsFile,
+                    std::function<void(std::string const &)> const & handler)
+{
+  ASSERT(handler != nullptr, ());
+  try
+  {
+    std::string fileData;
+    ReaderPtr<Reader>(GetPlatform().GetReader(iconsFile)).ReadAsString(fileData);
+    strings::Tokenize(fileData, "\n", [&handler](std::string const & str) {
+      if (!str.empty())
+        handler(str);
+    });
+  }
+  catch (Reader::Exception const & e)
+  {
+    LOG(LWARNING, ("Error reading file ", iconsFile, " : ", e.what()));
+  }
+}
+}  // namespace
+
+namespace local_ads
+{
+IconsInfo & IconsInfo::Instance()
+{
+  static IconsInfo iconsInfo;
+  return iconsInfo;
+}
+
+void IconsInfo::SetSourceFile(std::string const & fileName)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
```

Comment:
I mean, why not to do all heavy tasks as is, then take mutex and just assign m_fileName to fileName and swap m_icons with already filled icons?